client.go

```diff
@@ -628,8 +628,10 @@ type DialFunc func(addr string) (net.Conn, error)
 // Request argument passed to RetryIfFunc, if there are any request errors.
 type RetryIfFunc func(request *Request) bool
 
-// TransportFunc wraps every request/response.
-type TransportFunc func(*Request, *Response) error
+// RoundTripper wraps every request/response.
+type RoundTripper interface {
+	RoundTrip(hc *HostClient, req *Request, resp *Response) (retry bool, err error)
```

Comment:
It feels weird to have `*HostClient` in there, why is that needed? We're going to have people who try to implement the RoundTripper by modifying something and then trying to call `hc.Do` for example. Which will result in an infinite loop.

Reply:
> It feels weird to have *HostClient in there, why is that needed?

`RoundTripe` relies on `hc`'s fields and methods to create and release conns. If move these settings into the `Transport` struct, ay cause forward compatibility issues.

> We're going to have people who try to implement the RoundTripper by modifying something and then trying to call hc.Do for example. Which will result in an infinite loop.

The `TransportWrapper` testcase shows how to implement a coustom `RoundTripper` like go `net` sdk.